src/unicore.cpp

```diff
@@ -0,0 +1,149 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "unicore.h"
+#include "crc.h"
+#include <cstdio>
+#include <stdlib.h>
+#include <string.h>
+
+UnicoreParser::Result UnicoreParser::parseChar(char c)
+{
+	switch (_state) {
+	case State::Uninit:
+		if (c == '#') {
+			_state = State::GotHashtag;
+			_buffer_pos = 0;
+			_buffer_crc_pos = 0;
+		}
+
+		break;
+
+	case State::GotHashtag:
+		if (c == '*') {
+			_state = State::GotStar;
+
+			// Make sure buffer is zero terminated.
+			_buffer[_buffer_pos] = '\0';
+
+		} else {
+			if (_buffer_pos >= sizeof(_buffer)) {
+				_state = State::Uninit;
+				return Result::None;
+			}
+
+			_buffer[_buffer_pos++] = c;
+		}
+
+		break;
+
+	case State::GotStar:
+		_buffer_crc[_buffer_crc_pos++] = c;
+
+		if (_buffer_crc_pos == 8) {
+
+			// Make sure the CRC buffer is zero terminated.
+			_buffer_crc[_buffer_crc_pos] = '\0';
+
+			if (!crcCorrect()) {
+				return Result::WrongCrc;
+			}
+
+			if (isHeading()) {
+				if (extractHeading()) {
+					return Result::GotHeading;
+
+				} else {
+					return Result::WrongStructure;
+				}
+
+			} else {
+				return Result::UnknownSentence;
+			}
+		}
+
+		break;
+	}
+
+	return Result::None;
+}
+
+bool UnicoreParser::crcCorrect() const
+{
+	const uint32_t crc_calculated = calculateBlockCRC32(_buffer_pos, (uint8_t *)_buffer, 0);
+	const uint32_t crc_parsed = (uint32_t)strtoul(_buffer_crc, NULL, 16);
+	return (crc_calculated == crc_parsed);
+}
+
+bool UnicoreParser::isHeading() const
+{
+	const char header[] = "UNIHEADINGA";
+
+	if (strncmp(header, _buffer, strlen(header)) != 0) {
+		return false;
+	}
+
+	return true;
+}
+
+bool UnicoreParser::extractHeading()
+{
+	// The basline starts after ;,, and then follows the heading.
+
+	// Skip to ;
+	char *ptr = strchr(_buffer, ';');
+
+	if (ptr == NULL) {
+		return false;
+	}
+
+	ptr = strtok(ptr, ",");
+
+	unsigned i = 0;
+
+	while (ptr != NULL) {
+		ptr = strtok(NULL, ",");
+
+		if (i == 1) {
+			_heading.baseline_m = strtof(ptr, NULL);
+
+		} else if (i == 2) {
+			_heading.heading_deg = strtof(ptr, NULL);
+			break;
+		}
+
+		++i;
+	}
+
+	return true;
```

Comment:
`return i >= 5;`

Reply:
Actually, I was going to just early return when i == 5 was parsed.